test/call.c

```diff
@@ -158,6 +213,78 @@ static bool agent_audio_video_estab(const struct agent *ag)
 }
 
 
+static void process_rules(struct agent *ag, enum ua_event ev, const char *prm)
+{
+	struct fixture *f = ag->fix;
+	struct le *le;
+
+	LIST_FOREACH(&f->rules, le) {
+		struct cancel_rule *rule = le->data;
+
+		if (ev != rule->ev)
+			continue;
+
+		if (str_isset(rule->prm) &&
+		    str_casecmp(prm, rule->prm)) {
+			info("test: event %s prm=%s (expected %s)\n",
+			     uag_event_str(ev), prm, rule->prm);
+			continue;
+		}
+
+		if (rule->ua &&
+		    ag->ua != rule->ua) {
+			info("test: event %s ua=[%s] (expected [%s]\n",
+			     uag_event_str(ev),
+			     account_aor(ua_account(ag->ua)),
+			     account_aor(ua_account(rule->ua)));
+			continue;
+		}
+
+		if (rule->n_incoming &&
+		    ag->n_incoming != rule->n_incoming) {
```

Comment:
is the rule to compare exact number of events ?

or can the number of events be equal to or higher ?

remember that the counters should be compared after re_main is done.

Reply:
For `n_incoming` and the others currently proposed it's better to compare exact number. Later we will add counters for audio/video frames for which we specify lower bounds.

The counters are updated in the `event_handler` before the rules are evaluated. If all specified rules are met, then `re_cancel()` is called and after `re_main` is done further ASSERT_ checks can be done. Then optimally more API calls will trigger further actions and `re_main` is called again.